JavaScript/chapter02/p01_remove_dups/avc278.js

```diff
@@ -0,0 +1,101 @@
+// Remove Dups: Write code to remove duplicates from an unsorted linked list.
+// Follow Up: How would you solve this problem if a temporary buffer is not allowed?
+
+const assert = require("assert");
+
+class LinkedListNode {
+  constructor(val, next) {
+    this.val = val === undefined ? 0 : val;
+    this.next = next === undefined ? null : next;
+  }
+}
+
+/**
+ * Removes duplicates from a linked list
+ * @param  {LinkedList} LL input linked list to mutate
+ * @return {null}
+ *
+ * Keep a temporary set for storing values we pass through, of maximum size N where N is the number of nodes in the
+ * input linked list in the worst case. This requires an additional O(N) space. We store two pointers as we traverse
+ * through the linked list, skipping nodes with values already in our set.
+ * Runtime: O(N)
+ * Space:   O(N)
+ *
+ */
+const removeDups1 = (LL) => {
+  const seen = new Set();
+  let currNode = LL;
+  let next = LL.next;
+  seen.add(currNode.val);
+  while (next !== null) {
```

Comment:
_maybe_ this can be expressed as a `for` loop:

```javascript
for (let nextNode = currNode.next; nextNode.next !== null; nextNode = nextNode.next) {
```

same in L60 (and maybe also L58)

Reply:
Hmm yeah that works.

I'm having a hard time, though, seeing what the benefit of doing it that way is, as opposed to just using a regular while loop.
Is the purpose here to clean up the variable initialization, and increment-related code, so as to only have the duplication-related code in the loop?

Reply:
yup! programming is all about building using patterns (where they make sense). by writing the loop in this way, it's more clear that a "normal" traversal is being done instead of having to read a bit more of the code to realize that same information.

of course sometimes coercing the code to fit a pattern is detrimental, so it's not a universally-applicable rule :/

if this were production code, making a `LinkedList` that implements the [iterator protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) might be even better so that it could use a normal [`for ... of`]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop. but here it may be way overkill.